src/Microsoft.AspNet.Server.Kestrel/Http/PathNormalizer.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+
+namespace Microsoft.AspNet.Server.Kestrel.Http
+{
+    public static class PathNormalizer
+    {
+        public static bool NeedsNormalization(string path)
+        {
+            return !path.IsNormalized(NormalizationForm.FormC) || ContainsDotSegments(path);
+        }
+
+        public static string Normalize(string path)
+        {
+            if (path.IndexOf('/') > -1)
+            {
+                var normalizedChars = new char[path.Length];
+                var normalizedIndex = normalizedChars.Length;
+                var pathIndex = path.Length - 1;
+                var skipSegments = 0;
+
+                while (pathIndex >= 0)
+                {
+                    if (pathIndex >= 2 && path[pathIndex] == '.' && path[pathIndex - 1] == '.' && path[pathIndex - 2] == '/')
+                    {
+                        if (normalizedIndex == normalizedChars.Length || normalizedChars[normalizedIndex] != '/')
+                        {
+                            normalizedChars[--normalizedIndex] = '/';
+                        }
+
+                        skipSegments++;
+                        pathIndex -= 3;
+                    }
+                    else if (pathIndex >= 1 && path[pathIndex] == '.' && path[pathIndex - 1] == '/')
+                    {
+                        pathIndex -= 2;
+                    }
+                    else
+                    {
+                        while (pathIndex >= 0)
+                        {
+                            var lastChar = path[pathIndex];
+
+                            if (skipSegments == 0)
+                            {
+                                normalizedChars[--normalizedIndex] = lastChar;
+                            }
+
+                            pathIndex--;
+
+                            if (lastChar == '/')
+                            {
+                                break;
+                            }
+                        }
+
+                        if (skipSegments > 0)
+                        {
+                            skipSegments--;
+                        }
+                    }
+                }
+
+                path = new string(normalizedChars, normalizedIndex, normalizedChars.Length - normalizedIndex);
+            }
+
+            if (!path.IsNormalized(NormalizationForm.FormC))
```

Comment:
I don't see any reason dot compression and unicode normalization have to be run at the same time. If you separate them then you can limit the unicode normalization to only run in the needDecode scenario.